litejpeg/core/huffman/huffmancore.py

```diff
@@ -0,0 +1,487 @@
+        first_rle_word = Signal(1)
+        # To provide delay in the IDLE state for synchronization.
+        delay = Signal(2)
+        # Decide wheather IDLE state is ready to work.
+        ready_one = Signal(1)
+
+        # Keep track of the data to come out.
+        bit_ptr = Signal(5)
+        # Number of blocks to go as output.
+        num_fifo_wrs = Signal(2)
+
+        # High when the data is ready to go out.
+        hfw_running = Signal(1)
+        # Keep track of the block to come out.
+        fifo_wrt_cnt = Signal(2)
+
+        # To get the data out of the ROM.
+        # vlc_dc, vlc_dc_size:
+        #     Get the data and number of bits out to the DC ROM.
+        # vlc_ac, vlc_ac_size :
+        #     Get the data and number of bits out to the AC ROM.
+        vlc_dc = Signal(16)
+        vlc_dc_size = Signal(5)
+        vlc_ac = Signal(16)
+        vlc_ac_size = Signal(5)
+        # Decide wheather the data_out is DC or AC
+        # depending on the first_rle_word.
+        vlc_d = Array(Signal() for i in range(16))
+        vlc_size_d = Signal(5)
+
+        # Get the output from the amplitude and number of bits required
+        # to store the amplitude.
+        # Providing delay to the output for providing synchronization.
+        vli_ext = Array(Signal() for i in range(16))
+        vli_ext_next = Array(Signal() for i in range(16))
+        vli_ext_next_next = Array(Signal() for i in range(16))
+        vli_ext_next_next_next = Array(Signal() for i in range(16))
+        vli_ext_size = Signal(5)
+        vli_ext_size_next = Signal(5)
+        vli_ext_size_next_next = Signal(5)
+        vli_ext_size_next_next_next = Signal(5)
+
+        # For giving output to the Padding state.
+        pad_byte = Signal(8)
+        # For starting the PADDING state
+        pad_reg = Signal(1)
+
+        # For extracting the information out of the HuffmanEncoder Tables.
+        self.comb += [
+
+         self.dc_rom.address.eq(self.size),
+         vlc_dc_size.eq(self.dc_rom.data_out_size),
+         vlc_dc.eq(self.dc_rom.data_out_code),
+         self.ac_rom.address1.eq(self.size),
+         self.ac_rom.address2.eq(self.runlength),
+         vlc_ac_size.eq(self.ac_rom.data_out_size),
+         vlc_ac.eq(self.ac_rom.data_out_code)
+
+        ]
+
+        # For extracting the information for the Amplitude
+        # and size of the amplitude.
+        self.comb += [
+         [If(j < vli_ext_size,
+             vli_ext[j].eq(self.amplitude[j]))
+             for j in range(12)],
+         vli_ext_size.eq(self.size)
+        ]
+
+        self.sync += [
+
+         # MUX for selecting between AC and DC ROM.
+         If(first_rle_word,
+            [vlc_d[j].eq(vlc_dc[j]) for j in range(16)],
+            vlc_size_d.eq(vlc_dc_size))
+         .Else(
+            [vlc_d[j].eq(vlc_ac[j]) for j in range(16)],
+            vlc_size_d.eq(vlc_ac_size))
+
+        ]
+
+        # Calculating the number of blocks to go as output.
```

Comment:
The number of spaces for indenting in this file is wrong. You should always be using 4 spaces.

Reply:
Done